src/analyzer/protocol/ssl/ssl-dtls-protocol.pac

```diff
@@ -96,7 +96,7 @@ type UnknownRecord(rec: SSLRecord) = record {
 };
 
 type CiphertextRecord(rec: SSLRecord) = record {
-	cont : bytestring &restofdata &transient;
+	cont : bytestring &restofdata;
```

Comment:
Is there a risk of this changing performance substantially?

Reply:
So - I was curious about this, since I never looked before what transient actually does to binpac generated code. It turns out that the only difference in code seems to be that the type of `cont` changes from `bytestring` to `const_bytestream`, and that the accessor is removed.

A `const_bytestring` in turn is a `const_datastring<uint8>`, whereas a `bytestring` is a `datastring<uint8>`.

The difference between a const_datastring and a datatstring seems to be, that a datastring creates a copy of the data that it is initialized with, if the data that it is initialized with is const. If the data that it is initialized with is non-const, no copy is created, so the overhead of it is basically the same as of const_datastream (minus the fact that a free operation is called at the end, which is a no-op in this case).

In our case, there source data is indeed marked as const - so the difference is that one additional memcpy of all SSL data will be done. I admit that this seems not ideal. But - it turns out that this also means that my understanding of the meaning of &transient in binpac was completely wrong. It seems to be completely ok to access &transient members of a binpac parsed class, as long as you do it immediately (so - in &let fields that are attached to the class), and not at a later point of time when the pointers that are passed to ::Parse might no longer be available.

Thus, I actually expect that we could convert many (or even most?) of the occurrences of bytestring in zeek parsing to &transient, without any loss of functionality.

In any case - I am changing the code in this case to work with &transient; thus there is no significant speed impact and no copyingof data - the only thing left that has to happen is a check to see if we are decrypting data.

Reply:
Thanks for digging in, makes sense.